server/api/router.go

```diff
@@ -188,5 +199,77 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
 	// Deprecated
 	rootRouter.HandleFunc("/ping", func(w http.ResponseWriter, r *http.Request) {}).Methods("GET")
 
-	return rootRouter
+	if svr.GetConfig().EnableConfigManager {
+		f := func() {
+			// TODO: support DELETE
+			componentRouter := apiRouter.PathPrefix("/component").Methods("POST", "GET").Subrouter()
+			CustomForwardResponseOption := func(ctx context.Context, w http.ResponseWriter, pm proto.Message) error {
+				if _, ok := pm.(*configpb.GetResponse); ok {
+					str := pm.(*configpb.GetResponse).GetConfig()
+					w.Write([]byte(str))
+				}
+				return nil
+			}
+			gwmux := runtime.NewServeMux(
+				runtime.WithForwardResponseOption(CustomForwardResponseOption),
+				runtime.WithMarshalerOption("application/json", &runtime.JSONPb{OrigName: true}),
+				runtime.WithMarshalerOption("application/toml", &customMarshaler{}),
+			)
+			UnaryClientInterceptor := func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+				invoker(ctx, method, req, reply, cc, opts...)
+				var errMsg string
+				switch method {
+				case "/configpb.Config/Update":
+					errMsg = reply.(*configpb.UpdateResponse).GetStatus().GetMessage()
+				case "/configpb.Config/Get":
+					errMsg = reply.(*configpb.GetResponse).GetStatus().GetMessage()
+				}
+				if errMsg != "" {
+					return errors.New(errMsg)
+				}
+				return nil
+			}
+			tlsCfg, err := svr.GetSecurityConfig().ToTLSConfig()
+			if err != nil {
+				log.Error("fail to use TLS, use insecure instead", zap.Error(err))
+			}
+			opt := grpc.WithInsecure()
+			if tlsCfg != nil {
+				creds := credentials.NewTLS(tlsCfg)
+				opt = grpc.WithTransportCredentials(creds)
+			}
+			opts := []grpc.DialOption{opt, grpc.WithUnaryInterceptor(UnaryClientInterceptor)}
+			addr := svr.GetAddr()
+			u, err := url.Parse(addr)
+			if err != nil {
+				log.Error("failed to parse url", zap.Error(err))
+				return
+			}
+			err = configpb.RegisterConfigHandlerFromEndpoint(ctx, gwmux, u.Host+u.Path, opts)
+			if err != nil {
+				log.Error("fail to register grpc gateway", zap.Error(err))
+				return
+			}
+
+			componentRouter.Handle("", gwmux).Methods("POST")
+			componentRouter.Handle("/{component_id}", gwmux).Methods("GET")
+			componentRouter.Use(newComponentMiddleware(svr).Middleware)
+		}
+		return rootRouter, f
+	}
+	return rootRouter, nil
 }
+
+type customMarshaler struct{}
+
+func (c *customMarshaler) Marshal(v interface{}) ([]byte, error) {
```

Comment:
Why are they all empty implementations?

Reply:
Because we need to implement Marshaler interface here